### PR TITLE
2a. refactor(rpc): Add the ChainTip and Network to RpcImpl

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We need to estimate the chain height for these RPCs:
- #3143

This PR adds the fields, but doesn't implement the RPC.

## Solution

- Add the ChainTip and Network to RpcImpl
- Simplify RPC version field using generics

## Review

This PR could cause merge conflicts, so I'm marking it as a high priority.

It is based on PR #3847.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Implement #3143.

### CODE REVIEW

File: zebra-rpc/src/methods.rs [hunk 1]
```diff
@@ -111,47 +113,66 @@ pub trait Rpc {
...
         zebra_state::Request,
         Response = zebra_state::Response,
         Error = zebra_state::BoxError,
     >,
+    Tip: ChainTip + Send + Sync,
 {
     /// Create a new instance of the RPC handler.
-    pub fn new(
-        app_version: String,
+    pub fn new<Version>(
```
Comment (upbqdn) [inline zebra-rpc/src/methods.rs:148]: This is pretty neat.

File: zebra-rpc/src/methods/tests/prop.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+//! Randomised property tests for RPC methods.
+
```
Comment (upbqdn) [inline zebra-rpc/src/methods/tests/prop.rs:2]: One of the reasons why I add these comments as well is that if the file doesn't have it, my syntax highlighting doesn't work for the imports in the file.

**Review by upbqdn — DISMISSED** [review]
All looks good.

File: zebra-rpc/src/methods.rs (hunk 1 above)
Comment (teor2345) [reply zebra-rpc/src/methods.rs:148]: Yeah using a `Version` generic makes testing a lot easier.

**Review by oxarbitrage — APPROVED** [review]
Already approved PR with rebase.

**Pull request merged**